peps/pep-9999.rst

```diff
@@ -0,0 +1,256 @@
+  attestations, those attestations would be located at
+  ``HolyGrail-1.0.tar.gz.attestations``.
+* The attestations served by the index **SHALL** be in JSON format, with the
+  same well-known keys as accepted by the upload endpoint.
+
+.. TODO: Represent the trusted publisher metadata here, somehow.
+
+JSON-based Simple API
+^^^^^^^^^^^^^^^^^^^^^
+
+* When an uploaded file has an attestation, the index **MAY** include an
+  ``attestations`` key in the ``file`` dictionary for that file.
+* The ``attestations`` key, when present, **MUST** be a JSON object with
+  the same well-known keys and values as accepted by the upload endpoint.
+
+In addition to the above, when an uploaded file was uploaded via
+`Trusted Publishing <https://docs.pypi.org/trusted-publishers/>`_:
+
+* The ``file`` dictionary **MAY** include a ``publisher`` key.
+* The ``publisher`` key, when present, **MUST** be a JSON object with a
+  representation of the file's Trusted Publisher configuration at the time
+  of upload.
+* The keys within the ``publisher`` object are specific to each
+  Trusted Publisher but include, at minimum, a ``type`` key uniquely identifying
+  the type of Trusted Publisher. For example, for a GitHub-based Trusted
+  Publisher, the ``type`` may be the string ``"GitHub"``.
+
+Security Implications
+=====================
+
+This PEP is "mechanical" in nature; it provides only the plumbing for future
+digital attestations on package indices, without specifying their concrete
+cryptographic details.
+
+As such, we do not identify any positive or negative security implications
+for this PEP.
+
+Appendix 1: Example Uploaded Attestations
+=========================================
+
+This appendix provides a fictional example of the ``attestations`` field
+submitted on file upload, with two fictional attestations (``publish`` and
+``timestamp``):
+
+.. code-block:: json
+
+    {
+        "publish": {
+            "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
+            "verificationMaterial": { /* omitted for brevity */ },
+            "messageSignature": {
+                "messageDigest": {
+                    "algorithm": "some-hash-algo",
+                    "digest": "digest-here"
+                },
+                "signature": "signature-here"
+            }
+        },
+        "timestamp": {
+            "cms": "some-long-blob-here"
+        }
+    }
+
+The payloads of these fictional attestations are purely illustrative.
+
+Appendix 2: Example Trusted Publisher Representation
+====================================================
+
+This appendix provides a fictional example of a ``publisher`` key within
+a :pep:`691` ``project.files[]`` listing:
+
+.. code-block:: json
+
+    "publisher": {
+        "type": "GitHub",
+        "repository_name": "HolyGrail",
+        "repository_owner": "octocat",
+        "repository_owner_id": "1",
+        "workflow_filename": "publish.yml",
+        "environment": null
```

Comment:
(Copying discussion from chat.)

Yeah, with this level of state there's no guarantee of workflow contents, unfortunately -- to get that, we would indeed need the `ref` claim from the OIDC token itself (which isn't an engineering problem, but isn't technically part of the Trusted Publisher configuration). Alternatively, for a subset of publishers using reusable GitHub workflows, we could make somewhat stronger workflow content guarantees by checking the reusable workflow instead. But that second option isn't actually supported by PyPI yet.

TL;DR yes, I think we could add some kind of `ref` field that would provide direct provenance for a specific source revision. But we'll need to think about _how_ we want to expose that, since it's not technically part of the Trusted Publisher configuration itself.

Reply:
Folding this into the resolution for https://github.com/python/peps/pull/3618#discussion_r1453718488